crates/cranelift/src/wasmfx/optimized.rs

```diff
@@ -1463,17 +1506,158 @@ pub(crate) fn typed_continuations_store_payloads<'a>(
     }
 }
 
-/// We only use this when we want to get the current continuation for the
-/// purpose of suspending. Thus, if there is no such continuation (because we
-/// are on the main stack), we trap with an unhandled tag error.
-pub(crate) fn typed_continuations_load_continuation_reference<'a>(
+/// This function generates code that searches for a handler for `tag_address`,
+/// which must be a `*mut VMTagDefinition`. The search walks up the chain of
+/// continuations beginning at `start`.
+///
+/// The returned values are:
+/// 1. The continuation whose *parent* (another continuation or the
+///    main stack) handles `tag_address`.
+/// 2. The index of the handler in the parent's HandlerList.
+///
+/// In pseudo-code, the generated code's behavior can be expressed as
+/// follows:
+///
+/// chain_link = start
+/// while !chain_link.is_main_stack() {
+///   contref = chain_link.get_contref()
+///   parent_link = contref.parent
+///   parent_csi = parent_link.get_common_stack_information();
+///   handlers = parent_csi.handlers;
+///   len = handlers.length;
+///   for index in 0..len {
+///     if handlers[index] == tag_address {
+///       goto on_match(contref, index)
+///     }
+///   }
+///   chain_link = parent_link
+/// }
+/// trap(unhandled_tag)
+///
+/// on_match(conref : VMContRef, handler_index : u32)
+/// ... execution continues here here ...
+///
+fn search_handler<'a>(
     env: &mut crate::func_environ::FuncEnvironment<'a>,
     builder: &mut FunctionBuilder,
-) -> ir::Value {
-    let vmctx = env.vmctx_val(&mut builder.cursor());
-    let vmctx = tc::VMContext::new(vmctx, env.pointer_type());
-    let active_stack_chain = vmctx.load_stack_chain(env, builder);
-    active_stack_chain.unwrap_continuation_or_trap(env, builder, crate::TRAP_UNHANDLED_TAG)
+    start: &tc::StackChain,
+    tag_address: ir::Value,
+) -> (ir::Value, ir::Value) {
+    let handle_link = builder.create_block();
+    let begin_search_handler_list = builder.create_block();
+    let try_index = builder.create_block();
+    let compare_tags = builder.create_block();
+    let on_match = builder.create_block();
+    let on_no_match = builder.create_block();
+
+    // Terminate previous block:
+    builder.ins().jump(handle_link, &start.to_raw_parts());
+
+    // Block handle_link
+    let chain_link = {
+        builder.append_block_param(handle_link, env.pointer_type());
+        builder.append_block_param(handle_link, env.pointer_type());
+        builder.switch_to_block(handle_link);
+
+        let raw_parts = builder.block_params(handle_link);
+        let chain_link =
+            tc::StackChain::from_raw_parts([raw_parts[0], raw_parts[1]], env.pointer_type());
+        let is_main_stack = chain_link.is_main_stack(env, builder);
```

Comment:
I really don't like this special case of the main stack. But we can fix that in another patch. Ideally, every stack is homogeneous.

Reply:
My comment was conceptual. It is not about this particular code. We shouldn't have a notion of a main stack at all, is my point.